architecture/2024-10-user-defined-types-authoring-experience-feature-spec.md

```diff
@@ -0,0 +1,259 @@
+                                description: 'Name of the queue'
+                                required: true
+                            host:
+                                type: 'string'
+                                description: 'Hostname of the messaging service'
+                                required: true        
+                            port:
+                                type: 'string'
+                                description: 'Port'
+                                required: true
+                            connectionString:
+                                type: 'string'
+                                description: 'Connection string to the messaging service' 
+                                required: true 
+                    capabilities: ["Recipe"]           
+    ....
+    ```
+
+ Now Deb register this resource type via the CLI
+
+    ```bash
+    rad resource-provider create -f Mycompany.Messaging.yaml
+    ```
+The boilerplate code for CRUDL operations for the resource type is generated automatically and the resource type is registered in Radius.
+
+Now on creation, the schema should be validated for correctness and completeness. Deb should be provided with error messages if the schema is incorrect or incomplete.
+
+E.g.:
+
+    ```bash
+    rad resource-type create -f Mycompany.Messaging.yaml
+    Error: The schema definition is incorrect. The required property 'port' is missing a type in the schema definition
+    ```
+
+**Error Validations**:
+1. Syntax errors in the schema definition
+1. Required inputs are missing in the schema definition
+1. Incomplete data types for the properties in the schema definition
+
+### Scenario 2: Deb updates the schema for the internal messaging service Plaid
+<!-- One or two sentence summary -->
+Updates to schema definitions can be classified into two categories: non-breaking changes and breaking changes. 
+
+| Change type | Description | Example |
+|-------------|-------------|---------|
+| Non-breaking changes | Changes that can be made to the schema definition without impacting existing applications that use the resource type. | Eg: Adding an optional property to the schema definition |
+| Breaking changes | Changes that are not backward/forward compatible. These changes are disruptive and often requires the users to update their applications to use the new schema definition. | Eg: Adding a required property to the schema definition |
+
+There are two key sub-scenarios to consider when Deb updates the schema for the internal messaging service Plaid
+
+#### Sub-scenario 1: Deb adds a optional property and updates the resource type Plaid.
+Deb wants to add an optional property `logAnalytics` to have log analytics enabled for the resource-type. This is considered to be a non-breaking change as it is an optional property with default value. Users should be able to use the new property without any impact across versions on their existing applications.
+
+```yaml
```

Comment:
A few comments on this, maybe we should huddle and close down the following.

`resource namespace` isn't a great property name, because it requires quotes. I'd suggest either `namespace` or `name`.

---

See `resource namespace` again. A property can't both have a string value and children. We need to nest the types under another property. Open to suggestions on the naming, I like being concise and saying `types`.

This needs to look like:

```yaml
namespace: 'MyCompany.Messaging`
types:
  'plaidResources':
    apiVersions: ...
```

---

`required: true` isn't how required properties work in OpenAPI. It's more complicated 😆 

Here's the spec: https://json-schema.org/draft/2020-12/json-schema-validation#name-required (have fun)

Here's an example:

```yaml
schema:
  type: 'object'
  properties:
    a:
      type: string 
    b:
      type: string 
    c:
      type: string 
  required: [ 'a', 'b' ]
```

Note that `required` is an array of property names, and is a sibling of `properties`.

There's a HUGE amount of complexity in OpenAPI/JSON Schema. We'll support a **restricted** subset of what's possible.  Do you want me to take a stab at **that** document?

Reply:
I will update the eg with names and types as the properties. It would be helpful if you can help with the restricted subset of whats possible? It could be something small we start with and expand as we identify requirements based on usage/adoption

Reply:
I can help with that .... after Kubecon 😆